File: hushh-webapp/lib/api/consent.ts
/**
 * Consent API Client
 * ==================
 * 
 * Client for interacting with the Hushh Consent API.
 */

const API_BASE = process.env.NEXT_PUBLIC_CONSENT_API_URL || "http://localhost:8000";

// ============================================================================
// Types
// ============================================================================

export interface ConsentToken {
  token: string;
  user_id: string;
  agent_id: string;
  scope: string;
  issued_at: number;
  expires_at: number;
}

export interface ConsentScope {
  value: string;
  name: string;
}

// ============================================================================
// API Functions
// ============================================================================

/**
 * Issue a new consent token for a user/agent/scope combination.
 */
export async function issueToken(
  userId: string,
  agentId: string,
  scope: string
): Promise<ConsentToken> {
  const response = await fetch(`${API_BASE}/api/consent/issue`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      user_id: userId,
      agent_id: agentId,
      scope: scope,
    }),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Failed to issue token");
  }

  return response.json();
}

/**
 * Validate a consent token.
 */
export async function validateToken(
  token: string,
  expectedScope?: string
): Promise<{ valid: boolean; error?: string; user_id?: string }> {
  const response = await fetch(`${API_BASE}/api/consent/validate`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      token,
      expected_scope: expectedScope,
    }),
  });

  return response.json();
}

/**
 * Revoke a consent token.
 */
export async function revokeToken(token: string): Promise<{ success: boolean }> {
  const response = await fetch(`${API_BASE}/api/consent/revoke`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ token }),
  });

  if (!response.ok) {
    throw new Error("Failed to revoke token");
  }

  return response.json();
}

/**
 * Get available consent scopes.
 */
export async function getScopes(): Promise<ConsentScope[]> {
  const response = await fetch(`${API_BASE}/api/consent/scopes`);
  const data = await response.json();
  return data.scopes;
}

File: consent-protocol/api/routes/kai/consent.py
# api/routes/kai/consent.py
"""
Kai Consent Endpoints

Handles Kai-specific consent grants for analysis operations.
"""

import logging
import uuid
from typing import Dict, List

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from hushh_mcp.consent.token import issue_token
from hushh_mcp.constants import ConsentScope
from hushh_mcp.services.consent_db import ConsentDBService

logger = logging.getLogger(__name__)

router = APIRouter()


# ============================================================================
# MODELS
# ============================================================================

class GrantConsentRequest(BaseModel):
    user_id: str
    scopes: List[str] = [
        "vault.read.risk_profile",
        "vault.write.decision",
        "agent.kai.analyze",
    ]


class GrantConsentResponse(BaseModel):
    consent_id: str
    tokens: Dict[str, str]
    expires_at: str


# ============================================================================
# ENDPOINTS
# ============================================================================

@router.post("/consent/grant", response_model=GrantConsentResponse)
async def grant_consent(request: GrantConsentRequest):
    """
    Grant consent for Kai data access.
    
    Stateless: Issues tokens for the requested user_id and scopes.
    Does not rely on a pre-existing session.
    """
    service = ConsentDBService()
    
    tokens = {}
    consent_id = f"kai_consent_{uuid.uuid4().hex[:16]}"
    last_token_issued = None
    
    for scope_str in request.scopes:
        try:
            scope = ConsentScope(scope_str)
            token = issue_token(
                user_id=request.user_id,
                agent_id="agent_kai",
                scope=scope
            )
            tokens[scope_str] = token.token
            last_token_issued = token
            
            # Log to consent_audit using service layer
            await service.insert_event(
                user_id=request.user_id,
                agent_id="agent_kai",
                scope=scope_str,
                action="CONSENT_GRANTED",  # Use standard action name
                token_id=token.token[:32],  # Store truncated token ID
                expires_at=token.expires_at,
                issued_at=token.issued_at
            )
            
        except Exception as e:
            logger.error(f"Failed to issue token for scope {scope_str}: {e}")
            raise HTTPException(
                status_code=400,
                detail=f"Invalid scope: {scope_str}"
            )
    
    if not tokens:
        raise HTTPException(status_code=400, detail="No valid scopes provided")
    
    logger.info(f"[Kai] Consent granted for user: {request.user_id}")
    
    return GrantConsentResponse(
        consent_id=consent_id,
        tokens=tokens,
        expires_at=str(last_token_issued.expires_at) if last_token_issued else ""
    )
